src/rcc.rs

```diff
@@ -1,26 +1,341 @@
+                100_000 =>    0b0000,
+                200_000 =>    0b0001,
+                400_000 =>    0b0010,
+                800_000 =>    0b0011,
+                1_000_000 =>  0b0100,
+                2_000_000 =>  0b0101,
+                4_000_000 =>  0b0110,
+                8_000_000 =>  0b0111,
+                16_000_000 => 0b1000,
+                24_000_000 => 0b1001,
+                32_000_000 => 0b1010,
+                48_000_000 => 0b1011,
+                _ => panic!("bad MSI speed value!"),
+            };
+            MediumSpeedInternalRC {
+                f: f,
+                auto_cal: a,
+                bits: bits,
+            }
+        }
+
+        pub fn bits(&self) -> u8 {
+            self.bits
+        }
+
+        pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+            rcc.cr.modify(|_, w| unsafe {
+                w.msirange()
+                    .bits(self.bits())
+                    .msirgsel()
+                    .set_bit()
+            });
+            while rcc.cr.read().msirdy().bit_is_clear() {}
+
+            if self.auto_cal {
+                // FIXME This... may not work? I'm not sure if I've got a board problem or using
+                // the LSE requires some precondition I'm missing. In either case, LSERDY is never
+                // set by the hardware, so auto_cal doesn't succeed.
+                rcc.apb1enr1.modify(|_,w| w.pwren().set_bit());
+
+                rcc.bdcr.modify(|_,w| w.lseon().clear_bit());
+                while rcc.bdcr.read().lserdy().bit_is_set() {}
+                rcc.bdcr.modify(|_,w| unsafe {
+                    w.lsedrv()
+                        .bits(0b11)
+                        .lseon()
+                        .set_bit()
+                });
+                while rcc.bdcr.read().lserdy().bit_is_clear() {}
+                rcc.cr.modify(|_,w| w.msipllen().set_bit());
+            }
+            (self.freq(), 0b00)
+        }
+    }
+
+    impl InputClock for MediumSpeedInternalRC {
+        fn freq(&self) -> u32 {
+            self.f
+        }
+    }
+
+    /// High-speed external 4-48 MHz oscillator
+    #[derive(Clone, Copy)]
+    pub struct HighSpeedExternalOSC(pub u32);
+    impl InputClock for HighSpeedExternalOSC {
+        fn freq(&self) -> u32 {
+            self.0
+        }
+    }
+
+    impl HighSpeedExternalOSC {
+        pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+            rcc.cr.modify(|_, w| w
+                          .hseon()
+                          .set_bit()
+                         );
+            while rcc.cr.read().hserdy().bit_is_clear() {}
+            (self.freq(), 0b10)
+        }
+    }
+
+    pub enum RtcClkSource {
+        None,
+        /// Internal 32 kHz RC
+        LSI,
+        /// External 32.768 kHz oscillator
+        LSE,
+        HSEDiv32,
+    }
+    impl RtcClkSource {
+        pub fn freq(&self, hse: Option<HighSpeedExternalOSC>) -> Result<u32, ()> {
```

Comment:
I think Option<u32> works better here.
Having None is not error, it is just has no value.

Also please don't forget to add documentation for all Public methods